### PR TITLE
check if model is paralizable

### PR DESCRIPTION
**Pull request by janvanrijn**

### CODE REVIEW

File: openml/flows/sklearn_converter.py [hunk 1]
```diff
@@ -536,6 +536,49 @@ def _serialize_cross_validator(o):
 
     return ret
 
+def model_single_core(model):
```
Comment (mfeurer) [inline openml/flows/sklearn_converter.py]: Maybe call this `_check_n_jobs`?

File: openml/flows/sklearn_converter.py [hunk 2]
```diff
@@ -536,6 +536,49 @@ def _serialize_cross_validator(o):
...
+        param_distributions = None
+        if isinstance(model, sklearn.model_selection.GridSearchCV):
+            param_distributions = model.param_grid
+        elif isinstance(model, sklearn.model_selection.RandomizedSearchCV):
+            param_distributions = model.param_distributions
+        else:
+            print('Warning! Using subclass BaseSearchCV other than ' \
+                  '{GridSearchCV, RandomizedSearchCV}. Should implement param check. ')
+            pass
+
```
Comment (mfeurer) [inline openml/flows/sklearn_converter.py]: One newline is enough ;)

File: openml/flows/sklearn_converter.py [hunk 3]
```diff
@@ -536,6 +536,49 @@ def _serialize_cross_validator(o):
...
+                # measure runtime in a fair way
+                if value != 1 or disallow_parameter:
+                    return False
+        return True
+
+    if not (isinstance(model, sklearn.base.BaseEstimator) or
+            isinstance(model, sklearn.model_selection._search.BaseSearchCV)):
+        raise ValueError('model should be BaseEstimator or BaseSearchCV')
+
+    # check if the njobs is not in the optimization trace
```
Comment (mfeurer) [inline openml/flows/sklearn_converter.py]: I think you want to say: `make sure that n_jobs is not in the parameter grid`.

File: openml/flows/sklearn_converter.py [hunk 4]
```diff
@@ -536,6 +536,49 @@ def _serialize_cross_validator(o):
...
+                  '{GridSearchCV, RandomizedSearchCV}. Should implement param check. ')
+            pass
+
+
+        if not check(param_distributions, True):
+            raise PyOpenMLError('openml-python should not be used to '
+                                'optimize the n_jobs parameter.')
+
+    # check the parameters for n_jobs
+    if check(model.get_params(), False) == False:
```
Comment (mfeurer) [inline openml/flows/sklearn_converter.py]: You coud do `return check(model.get_params(), False)`

File: openml/flows/__init__.py [hunk 5]
```diff
@@ -1,6 +1,6 @@
 from .flow import OpenMLFlow
-from .sklearn_converter import sklearn_to_flow, flow_to_sklearn
+from .sklearn_converter import sklearn_to_flow, flow_to_sklearn,  model_single_core
 from .functions import get_flow, list_flows, flow_exists
 
 __all__ = ['OpenMLFlow', 'create_flow_from_model', 'get_flow', 'list_flows',
-           'sklearn_to_flow', 'flow_to_sklearn', 'flow_exists']
+           'sklearn_to_flow', 'flow_to_sklearn', 'flow_exists', 'model_is_paralizable']
```
Comment (mfeurer) [inline openml/flows/__init__.py]: What is `model_is_paralizable`?

File: tests/test_flows/test_sklearn.py [hunk 6]
```diff
@@ -555,3 +556,32 @@ def test_illegal_parameter_names_featureunion(self):
...
+        multicore_bagging = sklearn.ensemble.BaggingClassifier(n_jobs=5)
+        illegal_param_dist = {"base__n_jobs": [-1, 0, 1] }
+        legal_param_dist = {"base__max_depth": [2, 3, 4]}
+
+        legal_models = [
+            sklearn.ensemble.RandomForestClassifier(),
+            sklearn.ensemble.RandomForestClassifier(n_jobs=5),
+            sklearn.ensemble.RandomForestClassifier(n_jobs=-1),
+            sklearn.pipeline.Pipeline(steps=[('bag', sklearn.ensemble.BaggingClassifier(n_jobs=1))]),
+            sklearn.pipeline.Pipeline(steps=[('bag', sklearn.ensemble.BaggingClassifier(n_jobs=5))]),
```
Comment (mfeurer) [inline tests/test_flows/test_sklearn.py]: Shouldn't that fail because `n_jobs > 1`?

File: tests/test_flows/test_sklearn.py [hunk 7]
```diff
@@ -555,3 +556,32 @@ def test_illegal_parameter_names_featureunion(self):
...
+            sklearn.pipeline.Pipeline(steps=[('bag', sklearn.ensemble.BaggingClassifier(n_jobs=-1))]),
```
Comment (mfeurer) [inline tests/test_flows/test_sklearn.py]: Shouldn't that fail because `n_jobs < 0`?

File: tests/test_flows/test_sklearn.py [hunk 8]
```diff
@@ -555,3 +556,32 @@ def test_illegal_parameter_names_featureunion(self):
...
+    def test_paralizable_check(self):
+        singlecore_bagging = sklearn.ensemble.BaggingClassifier()
+        multicore_bagging = sklearn.ensemble.BaggingClassifier(n_jobs=5)
+        illegal_param_dist = {"base__n_jobs": [-1, 0, 1] }
+        legal_param_dist = {"base__max_depth": [2, 3, 4]}
+
+        legal_models = [
+            sklearn.ensemble.RandomForestClassifier(),
+            sklearn.ensemble.RandomForestClassifier(n_jobs=5),
+            sklearn.ensemble.RandomForestClassifier(n_jobs=-1),
```
Comment (mfeurer) [inline tests/test_flows/test_sklearn.py]: Shouldn't that fail because `n_jobs < 0`?

File: tests/test_flows/test_sklearn.py [hunk 9]
```diff
@@ -555,3 +556,32 @@ def test_illegal_parameter_names_featureunion(self):
...
+
+    def test_paralizable_check(self):
+        singlecore_bagging = sklearn.ensemble.BaggingClassifier()
+        multicore_bagging = sklearn.ensemble.BaggingClassifier(n_jobs=5)
+        illegal_param_dist = {"base__n_jobs": [-1, 0, 1] }
+        legal_param_dist = {"base__max_depth": [2, 3, 4]}
+
+        legal_models = [
+            sklearn.ensemble.RandomForestClassifier(),
+            sklearn.ensemble.RandomForestClassifier(n_jobs=5),
```
Comment (mfeurer) [inline tests/test_flows/test_sklearn.py]: Shouldn't that fail because `n_jobs > 1`?

File: tests/test_flows/test_sklearn.py [hunk 10]
```diff
@@ -555,3 +556,32 @@ def test_illegal_parameter_names_featureunion(self):
...
+            sklearn.pipeline.Pipeline(steps=[('bag', sklearn.ensemble.BaggingClassifier(n_jobs=-1))]),
+            sklearn.model_selection.GridSearchCV(singlecore_bagging, legal_param_dist),
+            sklearn.model_selection.GridSearchCV(multicore_bagging, legal_param_dist)
+        ]
+        illegal_models = [
+            sklearn.model_selection.GridSearchCV(singlecore_bagging, illegal_param_dist),
+            sklearn.model_selection.GridSearchCV(multicore_bagging, illegal_param_dist)
+        ]
+
+        answers = [True, False, False, True, False, False, True, False]
```
Comment (mfeurer) [inline tests/test_flows/test_sklearn.py]: Ah, here it's obvious what you're doing. Maybe you can re-order the test to make it easier to comprehend (i.e. check the legal models before declaring the illegal models, and maybe giving the list a different name.

File: openml/flows/sklearn_converter.py (hunk 1 above)
Comment (janvanrijn) [inline openml/flows/sklearn_converter.py]: Sure.

File: openml/flows/sklearn_converter.py (hunk 3 above)
Comment (janvanrijn) [inline openml/flows/sklearn_converter.py]: k

File: openml/flows/sklearn_converter.py (hunk 2 above)
Comment (janvanrijn) [inline openml/flows/sklearn_converter.py]: K.

File: openml/flows/sklearn_converter.py (hunk 4 above)
Comment (janvanrijn) [inline openml/flows/sklearn_converter.py]: Sure.

File: openml/flows/__init__.py (hunk 5 above)
Comment (janvanrijn) [inline openml/flows/__init__.py]: My bad. Removed it

File: tests/test_flows/test_sklearn.py (hunk 9 above)
Comment (janvanrijn) [inline tests/test_flows/test_sklearn.py]: Yes, this is exactly what happens, and this is specified in the answers array. Legal means here 'will not raise an error'. i added comments to make this more clear

File: tests/test_flows/test_sklearn.py (hunk 8 above)
Comment (janvanrijn) [inline tests/test_flows/test_sklearn.py]: Same as previous

File: tests/test_flows/test_sklearn.py (hunk 6 above)
Comment (janvanrijn) [inline tests/test_flows/test_sklearn.py]: Same as previous

File: tests/test_flows/test_sklearn.py (hunk 7 above)
Comment (janvanrijn) [inline tests/test_flows/test_sklearn.py]: Same as previous

File: tests/test_flows/test_sklearn.py (hunk 10 above)
Comment (janvanrijn) [inline tests/test_flows/test_sklearn.py]: I added some comments

**Pull request merged**